File: Typecast/TCMaxpTable.m
//
//  TCMaxpTable.m
//  Typecast
//
//  Created by David Schweinsberg on 3/12/12.
//  Copyright (c) 2012 David Schweinsberg. All rights reserved.
//

#import "TCMaxpTable.h"
#import "Type_Designer-Swift.h"

@implementation TCMaxpTable

- (id)initWithDataInput:(TCDataInput *)dataInput directoryEntry:(TCDirectoryEntry *)entry
{
    self = [super init];
    if (self)
    {
        self.directoryEntry = [entry copy];
        _versionNumber = [dataInput readUInt32];

        // CFF fonts use version 0.5, TrueType fonts use version 1.0
        if (_versionNumber == 0x00005000)
        {
            _numGlyphs = [dataInput readUInt16];
        }
        else if (_versionNumber == 0x00010000)
        {
            _numGlyphs = [dataInput readUInt16];
            _maxPoints = [dataInput readUInt16];
            _maxContours = [dataInput readUInt16];
            _maxCompositePoints = [dataInput readUInt16];
            _maxCompositeContours = [dataInput readUInt16];
            _maxZones = [dataInput readUInt16];
            _maxTwilightPoints = [dataInput readUInt16];
            _maxStorage = [dataInput readUInt16];
            _maxFunctionDefs = [dataInput readUInt16];
            _maxInstructionDefs = [dataInput readUInt16];
            _maxStackElements = [dataInput readUInt16];
            _maxSizeOfInstructions = [dataInput readUInt16];
            _maxComponentElements = [dataInput readUInt16];
            _maxComponentDepth = [dataInput readUInt16];
        }
    }
    return self;
}

- (uint32_t)type
{
    return TCTable_maxp;
}

- (NSString *)description
{
    NSString *str = [NSString stringWithFormat:
                     @"'maxp' Table - Maximum Profile\n------------------------------"
                     @"\n        'maxp' version:         %x"
                     @"\n        numGlyphs:              %d",
                     _versionNumber,
                     _numGlyphs];

    if (_versionNumber == 0x00010000)
        str = [str stringByAppendingFormat:
               @"\n        maxPoints:              %d" //).append(maxPoints)
               @"\n        maxContours:            %d" //).append(maxContours)
               @"\n        maxCompositePoints:     %d" //).append(maxCompositePoints)
               @"\n        maxCompositeContours:   %d" //).append(maxCompositeContours)
               @"\n        maxZones:               %d" //).append(maxZones)
               @"\n        maxTwilightPoints:      %d" //).append(maxTwilightPoints)
               @"\n        maxStorage:             %d" //).append(maxStorage)
               @"\n        maxFunctionDefs:        %d" //).append(maxFunctionDefs)
               @"\n        maxInstructionDefs:     %d" //).append(maxInstructionDefs)
               @"\n        maxStackElements:       %d" //).append(maxStackElements)
               @"\n        maxSizeOfInstructions:  %d" //).append(maxSizeOfInstructions)
               @"\n        maxComponentElements:   %d" //).append(maxComponentElements)
               @"\n        maxComponentDepth:      %d",//).append(maxComponentDepth);
               _maxPoints,
               _maxContours,
               _maxCompositePoints,
               _maxCompositeContours,
               _maxZones,
               _maxTwilightPoints,
               _maxStorage,
               _maxFunctionDefs,
               _maxInstructionDefs,
               _maxStackElements,
               _maxSizeOfInstructions,
               _maxComponentElements,
               _maxComponentDepth];
    else
        str = [str stringByAppendingString:@"\n"];
    return str;
}

@end

File: Typecast/TCMaxpTable.swift
//
//  TCMaxpTable.swift
//  Type Designer
//
//  Created by David Schweinsberg on 8/5/17.
//  Copyright © 2017 David Schweinsberg. All rights reserved.
//

import Foundation
import IOUtils

class TCMaxpTable: TCTable, Codable {
  let versionNumber: UInt32
  let numGlyphs: Int
  let maxPoints: UInt16
  let maxContours: UInt16
  let maxCompositePoints: UInt16
  let maxCompositeContours: UInt16
  let maxZones: UInt16
  let maxTwilightPoints: UInt16
  let maxStorage: UInt16
  let maxFunctionDefs: UInt16
  let maxInstructionDefs: UInt16
  let maxStackElements: UInt16
  let maxSizeOfInstructions: UInt16
  let maxComponentElements: UInt16
  let maxComponentDepth: UInt16

  override init() {
    versionNumber = 0
    numGlyphs = 0
    maxPoints = 0
    maxContours = 0
    maxCompositePoints = 0
    maxCompositeContours = 0
    maxZones = 0
    maxTwilightPoints = 0
    maxStorage = 0
    maxFunctionDefs = 0
    maxInstructionDefs = 0
    maxStackElements = 0
    maxSizeOfInstructions = 0
    maxComponentElements = 0
    maxComponentDepth = 0
    super.init()
  }

  init(data: Data) {
    let dataInput = TCDataInput(data: data)
    versionNumber = dataInput.readUInt32()

    // CFF fonts use version 0.5, TrueType fonts use version 1.0
    if versionNumber == 0x00005000 {
      numGlyphs = Int(dataInput.readUInt16())
      maxPoints = 0
      maxContours = 0
      maxCompositePoints = 0
      maxCompositeContours = 0
      maxZones = 0
      maxTwilightPoints = 0
      maxStorage = 0
      maxFunctionDefs = 0
      maxInstructionDefs = 0
      maxStackElements = 0
      maxSizeOfInstructions = 0
      maxComponentElements = 0
      maxComponentDepth = 0
    } else if versionNumber == 0x00010000 {
      numGlyphs = Int(dataInput.readUInt16())
      maxPoints = dataInput.readUInt16()
      maxContours = dataInput.readUInt16()
      maxCompositePoints = dataInput.readUInt16()
      maxCompositeContours = dataInput.readUInt16()
      maxZones = dataInput.readUInt16()
      maxTwilightPoints = dataInput.readUInt16()
      maxStorage = dataInput.readUInt16()
      maxFunctionDefs = dataInput.readUInt16()
      maxInstructionDefs = dataInput.readUInt16()
      maxStackElements = dataInput.readUInt16()
      maxSizeOfInstructions = dataInput.readUInt16()
      maxComponentElements = dataInput.readUInt16()
      maxComponentDepth = dataInput.readUInt16()
    } else {
      numGlyphs = 0
      maxPoints = 0
      maxContours = 0
      maxCompositePoints = 0
      maxCompositeContours = 0
      maxZones = 0
      maxTwilightPoints = 0
      maxStorage = 0
      maxFunctionDefs = 0
      maxInstructionDefs = 0
      maxStackElements = 0
      maxSizeOfInstructions = 0
      maxComponentElements = 0
      maxComponentDepth = 0
    }
    super.init()
  }

  override class var tag: TCTable.Tag {
    get {
      return .maxp
    }
  }

  override var description: String {
    get {
      var str = String(format:
        "'maxp' Table - Maximum Profile\n------------------------------" +
        "\n        'maxp' version:         %x" +
        "\n        numGlyphs:              %d",
        versionNumber,
        numGlyphs)

      if versionNumber == 0x00010000 {
        str.append(String(format:
          "\n        maxPoints:              %d" +
          "\n        maxContours:            %d" +
          "\n        maxCompositePoints:     %d" +
          "\n        maxCompositeContours:   %d" +
          "\n        maxZones:               %d" +
          "\n        maxTwilightPoints:      %d" +
          "\n        maxStorage:             %d" +
          "\n        maxFunctionDefs:        %d" +
          "\n        maxInstructionDefs:     %d" +
          "\n        maxStackElements:       %d" +
          "\n        maxSizeOfInstructions:  %d" +
          "\n        maxComponentElements:   %d" +
          "\n        maxComponentDepth:      %d",
          maxPoints,
          maxContours,
          maxCompositePoints,
          maxCompositeContours,
          maxZones,
          maxTwilightPoints,
          maxStorage,
          maxFunctionDefs,
          maxInstructionDefs,
          maxStackElements,
          maxSizeOfInstructions,
          maxComponentElements,
          maxComponentDepth))
      } else {
        str.append("\n")
      }
      return str;
    }
  }
}
